File: answers/leetcode/ZigZag Conversion/ZigZag Conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """

        if 1 == numRows:
            return s
        rows = ['' for i in range(numRows)]
        index = -1
        step = 1
        for i in range(len(s)):
            index += step
            if index >= numRows:
                index -= 2
                step = -1
            elif index <= -1:
                index = 1
                step = 1
            rows[index] += s[i]
        return ''.join(rows)

File: answers/leetcode/ZigZag Conversion/ZigZag Conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if (numRows <= 1) {
            return s;
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; i + j < s.length(); j += numRows * 2 - 2) {
                sb.append(s.charAt(i + j));
                if (i > 0 && i + 1 < numRows && j + numRows * 2 - 2 - i < s.length()) {
                    sb.append(s.charAt(j + numRows * 2 - 2 - i));
                }
            }
        }
        return sb.toString();
    }
}